locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicLocationNetworkInfoCustomizer.java

```diff
@@ -0,0 +1,472 @@
+
+    /**
+     * Combines the given resolve options with the customiser's configuration to determine the
+     * best address and credential pair for management. In particular, if the resolve options
+     * allow it will check that the credential is actually valid for the address.
+     */
+    @Override
+    public ManagementAddressResolveResult resolve(
+            JcloudsLocation location, NodeMetadata node, ConfigBag config, ManagementAddressResolveOptions options) {
+        LOG.debug("{} resolving management parameters for {}, node={}, config={}, options={}",
+                new Object[]{this, location, node, config, options});
+        Stopwatch timer = Stopwatch.createStarted();
+        // Should only be null in tests.
+        final Entity contextEntity = getContextEntity(config);
+        if (shouldPublishNetworks() && options.publishNetworkSensors() && contextEntity != null) {
+            publishNetworks(node, contextEntity);
+        }
+        HostAndPort hapChoice = null;
+        LoginCredentials credChoice = null;
+
+        Iterable<HostAndPort> managementCandidates = getManagementCandidates(location, node, config, options);
+        Iterable<LoginCredentials> credentialCandidates = getCredentialCandidates(location, node, options, config);
+
+        // Try each pair of address and credential until one succeeds.
+        if (options.expectReachable() && options.pollForFirstReachableAddress() && shouldCheckCredentials()) {
+            for (HostAndPort hap : managementCandidates) {
+                for (LoginCredentials cred : credentialCandidates) {
+                    LOG.trace("Testing host={} with credential={}", hap, cred);
+                    if (checkCredential(location, hap, cred, config, options.isWindows())) {
+                        hapChoice = hap;
+                        credChoice = cred;
+                        break;
+                    }
+                }
+                if (hapChoice != null) break;
+            }
+        }
+
+        if (hapChoice == null) {
+            LOG.trace("Choosing first management candidate given node={} and mode={}", node, getMode());
+            hapChoice = Iterables.getFirst(managementCandidates, null);
+        }
+        if (hapChoice == null) {
+            LOG.trace("Choosing first address of node={} in mode={}", node, getMode());
+            final Iterator<String> hit = getNodeAddressesWithMode(node).iterator();
+            if (hit.hasNext()) HostAndPort.fromHost(hit.next());
+        }
+        if (hapChoice == null) {
+            throw new IllegalStateException("Exhausted all options when determining address for " + location);
+        }
+
+        if (credChoice == null) {
+            credChoice = Iterables.getFirst(credentialCandidates, null);
+            if (credChoice == null) {
+                throw new IllegalStateException("Exhausted all options when determining credential for " + location);
```

Comment:
A more appropriate message would be `No credentials configured for location`. Since `credentialCandidates` is empty.